gk/main.c

```diff
@@ -157,28 +158,89 @@ priority_from_delta_time(uint64_t present, uint64_t past)
 	return integer_log_base_2(delta_time);
 }
 
+static int
+get_grantor_id(struct gk_lpm *ltbl, struct ip_flow *flow)
+{
+	int ret;
+	if (flow->proto == ETHER_TYPE_IPv4)
+		ret = lpm_lookup_ipv4(ltbl->lpm, flow->f.v4.dst);
+	else if (flow->proto == ETHER_TYPE_IPv6)
+		ret = lpm_lookup_ipv6(ltbl->lpm6, flow->f.v6.dst);
+	else
+		rte_panic("Unexpected condition at %s: unknown flow type %hu\n",
+			__func__, flow->proto);
+
+	return ret;
+}
+
+static struct gk_policy *
+get_gk_policy(struct gk_lpm *ltbl, struct ip_flow *flow)
+{
+	int gt_id = get_grantor_id(ltbl, flow);
+	return gt_id < 0 ? NULL : &ltbl->policy_tbl[gt_id];
+}
+
 static inline void
-initialize_flow_entry(struct flow_entry *fe, struct ip_flow *flow)
+initialize_flow_entry(struct flow_entry *fe,
+	struct ip_flow *flow, struct gk_lpm *ltbl)
 {
+	int gt_id;
+
 	rte_memcpy(&fe->flow, flow, sizeof(*flow));
 
 	fe->state = GK_REQUEST;
 	fe->u.request.last_packet_seen_at = rte_rdtsc();
 	fe->u.request.last_priority = START_PRIORITY;
 	fe->u.request.allowance = START_ALLOWANCE - 1;
-	/* TODO Grantor ID comes from LPM lookup. */
-	fe->u.request.grantor_id = 0;
+
+	/* Grantor ID comes from LPM lookup. */
+	gt_id = get_grantor_id(ltbl, &fe->flow);
+	if (gt_id < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: initialize flow entry error in %s\n",
+			__func__);
+		return;
+	}
+
+	/*
+	 * TODO The grantor id is being saved without adding a reference
+	 * to grantor entry. If the grantor entry is removed later by
+	 * the dynamic config block, the code will break.
+	 * We'll need to adopt RCU in the GK blocks.
+	 */
+	fe->u.request.grantor_id = gt_id;
 }
 
 static inline void
-reinitialize_flow_entry(struct flow_entry *fe, uint64_t now)
+reinitialize_flow_entry(
+	struct flow_entry *fe, struct gk_lpm *ltbl, uint64_t now)
 {
+	int gt_id = -1;
+
+	if (fe->state == GK_REQUEST)
+		gt_id = fe->u.request.grantor_id;
+	else if (fe->state == GK_GRANTED)
+		gt_id = fe->u.granted.grantor_id;
+
 	fe->state = GK_REQUEST;
 	fe->u.request.last_packet_seen_at = now;
 	fe->u.request.last_priority = START_PRIORITY;
 	fe->u.request.allowance = START_ALLOWANCE - 1;
-	/* TODO Grantor ID comes from LPM lookup. */
-	fe->u.request.grantor_id = 0;
+
+	if (gt_id >= 0)
+		fe->u.request.grantor_id = gt_id;
+	else {
+		/* Grantor ID comes from LPM lookup. */
+		gt_id = get_grantor_id(ltbl, &fe->flow);
+		if (gt_id < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: initialize flow entry error in %s\n",
+				__func__);
+			return;
```

Comment:
Similarly to the issue in `initialize_flow_entry()`, how can the caller know that something went wrong to do something about?

Here my suggestion of solution is a little trickier: adding grantor id to the declined state of the flow. The grantor id is known at both other states, so one can just copy the grantor id when changing from another state to the declined state. This will enable `reinitialize_flow_entry()` to always know the grantor ID without a new lookup.